sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala

```diff
@@ -67,68 +70,74 @@ case class BroadcastExchangeExec(
 
   @transient
   private lazy val relationFuture: Future[broadcast.Broadcast[Any]] = {
-    // broadcastFuture is used in "doExecute". Therefore we can get the execution id correctly here.
+    // relationFuture is used in "doExecute". Therefore we can get the execution id correctly here.
     val executionId = sparkContext.getLocalProperty(SQLExecution.EXECUTION_ID_KEY)
-    Future {
-      // This will run in another thread. Set the execution id so that we can connect these jobs
-      // with the correct execution.
-      SQLExecution.withExecutionId(sqlContext.sparkSession, executionId) {
-        try {
-          val beforeCollect = System.nanoTime()
-          // Use executeCollect/executeCollectIterator to avoid conversion to Scala types
-          val (numRows, input) = child.executeCollectIterator()
-          if (numRows >= 512000000) {
-            throw new SparkException(
-              s"Cannot broadcast the table with 512 million or more rows: $numRows rows")
-          }
-
-          val beforeBuild = System.nanoTime()
-          longMetric("collectTime") += NANOSECONDS.toMillis(beforeBuild - beforeCollect)
-
-          // Construct the relation.
-          val relation = mode.transform(input, Some(numRows))
-
-          val dataSize = relation match {
-            case map: HashedRelation =>
-              map.estimatedSize
-            case arr: Array[InternalRow] =>
-              arr.map(_.asInstanceOf[UnsafeRow].getSizeInBytes.toLong).sum
-            case _ =>
-              throw new SparkException("[BUG] BroadcastMode.transform returned unexpected type: " +
-                  relation.getClass.getName)
+    val task = new Callable[broadcast.Broadcast[Any]]() {
+      override def call(): broadcast.Broadcast[Any] = {
+        // This will run in another thread. Set the execution id so that we can connect these jobs
+        // with the correct execution.
+        SQLExecution.withExecutionId(sqlContext.sparkSession, executionId) {
+          try {
+            sparkContext.setJobGroup(runId.toString, s"broadcast exchange (runId $runId)",
```

Comment:
let's add a comment to explain why we set up a job group here. There is no other public API that can cancel a specific job AFAIK.

Reply:
Just curious, why can't we run this collect & broadcast in the same thread as the the query planning (a.k.a. main thread) so that when a job group is cancelled, these broadcast jobs can be cancelled as a whole? To me it looks like we are already awaiting the future whatsoever.

Reply:
That sounds like a good idea. We should only set the job group if there is no one outside.

Reply:
Wouldn't that cancelling the broadcast job cause the outer main job to cancel?

Reply:
> We should only set the job group if there is no one outside.

and I guess it would be a partial fix?

Reply:
I rechecked the code, we will cancel the broadcast if a timeout is hit, and we shouldn't cancel the main job there. We need a way to only cancel the jobs of the broadcast.

Reply:
seems like the problem is, we didn't set the job group id back to its original value after broadcast is done?

Reply:
Can we somehow encode the fact that this RDD is a broadcast job in the jobid/jobgroupid?

Reply:
I think that's hacky though. To fix it properly, we should either have multiple job group ids or job group hierarchy.

Reply:
I agree that we should have hierarchical job group.

Reply:
My idea is that since cancelling the broadcast job is an optimization method, and the UI is also user-oriented, it is better to add a configuration item, such as: 
`spark.broadcast.job.cancelling.enabled`
 When it is true, it supports cancelling the broadcast job. The default is false. By default,  if the user sets up job group and job description, he will not be surprised by the UI. If the user wants to optimize the broadcast job, he will not be surprised by the UI too. We can make it clear in the configuration doc. Mutiple job group can also achieve the goal, but I am worried that it will be more disruptive to program development. Some old users may not know how to set up mutiple job group without viewing the source code.In addition, multiple job group will bring more workload and higher error probability.

Reply:
It's always better to have fewer configs if possible. And I don't think we can override the job group id here if the config is true, as this is used to cancel broadcast after timeout.

Reply:
@HyukjinKwon  @jiangxb1987 @yeshengm What‘s your opinion of my idea?

Reply:
I am discussing about multiple job group support which will fundamentally fix all these problems. This is actually a general problem that's not speicfic to SQL broadcast here only.

Reply:
@HyukjinKwon Could you please tell me where you are discussing? I also want to make a little contribution.

Reply:
sorry I am discussing offline first. I will send out an email or JIRA soon.